mllib/src/main/scala/org/apache/spark/mllib/clustering/DpMeans.scala

```diff
@@ -0,0 +1,247 @@
+  def getConvergenceTol: Double = convergenceTol
+
+  /** Set the maximum number of iterations. Default: 20 */
+  def setMaxIterations(maxIterations: Int): this.type = {
+    this.maxIterations = maxIterations
+    this
+  }
+
+  /** Return the maximum number of iterations. */
+  def getMaxIterations: Int = maxIterations
+
+  /**
+   * Perform DP means clustering
+   */
+  def run(data: RDD[Vector]): DpMeansModel = {
+    if (data.getStorageLevel == StorageLevel.NONE) {
+      logWarning("The input data is not directly cached, which may hurt performance if its"
+        + " parent RDDs are also uncached.")
+    }
+
+    // Compute squared norms and cache them.
+    val norms = data.map(Vectors.norm(_, 2.0))
+    norms.persist()
+    val zippedData = data.zip(norms).map {
+      case (v, norm) => new VectorWithNorm(v, norm)
+    }
+
+    // Implementation of DP means algorithm.
+    var iteration = 0
+    var covered = false
+    var converged = false
+    var localCenters = Array.empty[VectorWithNorm]
+    val globalCenters = ArrayBuffer.empty[VectorWithNorm]
+
+    // Execute clustering until the maximum number of iterations is reached
+    // or the cluster centers have converged.
+    while (iteration < maxIterations && !converged) {
+      type WeightedPoint = (Vector, Long)
+      def mergeClusters(x: WeightedPoint, y: WeightedPoint): WeightedPoint = {
+        axpy(1.0, x._1, y._1)
+        (y._1, x._2 + y._2)
+      }
+
+      // Loop until all data points are covered by some cluster center
+      do {
+        localCenters = zippedData.mapPartitions(h => DpMeans.cover(h, globalCenters, lambda))
+          .collect()
+        if (localCenters.isEmpty) {
+          covered = true
+        }
+        // Promote a local cluster center to a global center
+        else {
+          var newGlobalCenters = DpMeans.cover(localCenters.iterator,
+            ArrayBuffer.empty[VectorWithNorm], lambda)
+          globalCenters ++= newGlobalCenters
+        }
+      } while (covered == false)
+
+      // Find the sum and count of points belonging to each cluster
+      val clusterStat = zippedData.mapPartitions { points =>
+        val activeCenters = globalCenters
+        val k = activeCenters.length
+        val dims = activeCenters(0).vector.size
+
+        val sums = Array.fill(k)(Vectors.zeros(dims))
+        val counts = Array.fill(k)(0L)
+        val totalCost = Array.fill(k)(0.0D)
+
+        points.foreach { point =>
+          val (currentCenter, cost) = DpMeans.assignCluster(activeCenters, point)
+          totalCost(currentCenter) += cost
+          val currentSum = sums(currentCenter)
+          axpy(1.0, point.vector, currentSum)
+          counts(currentCenter) +=1
+        }
+
+        val result = for (i <- 0 until k) yield {
+          (i, (sums(i), counts(i)))
+        }
+        result.iterator
+      }.reduceByKey(mergeClusters).collectAsMap()
```

Comment:
We should use `aggregateByKey` instead of `reduceByKey` because `reduceByKey` shouldn't modify the items in place. It might be okay to define `mergeClusters` inline with pattern matching to improve readability.